facade-charm/src/charm.py

```diff
@@ -0,0 +1,192 @@
+        updated = []
+        for relation in to_update:
+            if self._update_relation(relation, clear=clear):
+                updated.append(relation.name)
+        return updated
+
+    def _update_relation(self, relation: ops.Relation,
+                         clear=False, replace=False):
+        changed = False
+
+        app_databag = relation.data[self.app]
+        unit_databag = relation.data[self.unit]
+        if replace or clear:
+            if app_databag.keys():
+                app_databag.clear()
+                changed = True
+            if unit_databag.keys():
+                unit_databag.clear()
+                changed = True
+
+        if not clear:
+            app_data, unit_data = self._load_mock(relation.name)
+            if app_data and dict(app_databag) != app_data:
+                app_databag.update(app_data)
+                changed = True
+            if unit_data and dict(unit_databag) != unit_data:
+                unit_databag.update(unit_data)
+                changed = True
+        return changed
+
+    def _load_mock(self, endpoint: str):
+        mocks_root = Path(__file__).parent.parent / 'mocks'
+
+        if endpoint.startswith("provide"):
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
+        else:
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
+
+        if not pth.exists():
+            logger.warning(f"mock not found for {endpoint} ({pth})")
+            pth.touch()
+            self._write_mock(endpoint, {}, {})
+            return {}, {}
+
+        yml = yaml.safe_load(pth.read_text())
+        app_data = yml.get("app_data", {})
+        unit_data = yml.get("unit_data", {})
+        return app_data, unit_data
+
+    def _write_mock(self, endpoint: str,
+                    app_data: Optional[dict] = None,
+                    unit_data: Optional[dict] = None):
+        mocks_root = Path(__file__).parent.parent / 'mocks'
+
+        if endpoint.startswith("provide"):
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
+        else:
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
+
+        if not pth.exists():
+            logger.info(f"mock not found for {endpoint}")
+            pth.touch()
+
+        yml = yaml.safe_load(pth.read_text())
+
+        if app_data == {}:
+            _app_data = {}
+        else:
+            _app_data = yml.get("app_data") or {}
+            if app_data:
+                _app_data.update(app_data)
+
+        if unit_data == {}:
+            _unit_data = {}
+        else:
+            _unit_data = yml.get("unit_data") or {}
+            if unit_data:
+                _unit_data.update(unit_data)
+
+        logger.info(f"updating mock with {_app_data}, {_unit_data}")
+        pth.write_text(
+            yaml.safe_dump(
+                {
+                    "endpoint": endpoint,
+                    "app_data": _app_data,
+                    "unit_data": _unit_data}
```

Comment:
minor: consider pre-commit

Reply:
not everyone is used to it (unfortunately), it's something we should consider adopting project-wide, out of scope for this pr. Added a `tox fmt` env though.